import importlib
from itertools import groupby 
from telegram import Update
from telegram.ext import Updater, CommandHandler, CallbackContext
from telegram.ext import InlineQueryHandler,CallbackQueryHandler, ChosenInlineResultHandler
import motor.motor_asyncio
import urllib.request
import random
from shivu import application, sudo_users, collection 


async def get_next_sequence_number(sequence_name):
    sequence_collection = db.sequences
    sequence_document = await sequence_collection.find_one_and_update(
        {'_id': sequence_name}, 
        {'$inc': {'sequence_value': 1}}, 
        return_document=ReturnDocument.AFTER
    )
    if not sequence_document:
        await sequence_collection.insert_one({'_id': sequence_name, 'sequence_value': 0})
        return 0
    return sequence_document['sequence_value']

async def upload(update: Update, context: CallbackContext) -> None:
    if str(update.effective_user.id) not in sudo_users:
        await update.message.reply_text('You do not have permission to use this command.')
        return

    try:
        args = context.args
        if len(args) != 4:
            await update.message.reply_text('Incorrect format. Please use: /upload img_url Character-Name Anime-Name Rarity')
            return

        character_name = args[1].replace('-', ' ').title()
        anime = args[2].replace('-', ' ').title()

        try:
            urllib.request.urlopen(args[0])
        except:
            await update.message.reply_text('Invalid image URL.')
            return

        rarity_map = {1: "âšª Common", 2: "ðŸŸ£ Rare", 3: "ðŸŸ¡ Legendary", 4: "ðŸŸ¢ Medium"}
        try:
            rarity = rarity_map[int(args[3])]
        except KeyError:
            await update.message.reply_text('Invalid rarity. Please use 1, 2, 3, or 4.')
            return

        id = str(await get_next_sequence_number('character_id')).zfill(2)

        character = {
            'img_url': args[0],
            'name': character_name,
            'anime': anime,
            'rarity': rarity,
            'id': id
        }

        message = await context.bot.send_photo(
            chat_id='-1001915956222',
            photo=args[0],
            caption=f'<b>Character Name:</b> {character_name}\n<b>Anime Name:</b> {anime}\n<b>Rarity:</b> {rarity}\n<b>ID:</b> {id}\nAdded by <a href="tg://user?id={update.effective_user.id}">{update.effective_user.first_name}</a>',
            parse_mode='HTML'
        )

        character['message_id'] = message.message_id
        await collection.insert_one(character)

        await update.message.reply_text('DONE.. CHEACK IN @characters_database')
    except Exception as e:
        await update.message.reply_text(f'Unsuccessfully uploaded. Error: {str(e)}')
